File: src/test/java/application/cleanup/CleanupTest.java
package application.cleanup;

import application.AbstractApplicationTest;
import application.InitTestServiceDecorator;
import com.matag.cards.Cards;
import com.matag.game.cardinstance.CardInstance;
import com.matag.game.init.test.InitTestService;
import com.matag.game.status.GameStatus;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import static application.browser.BattlefieldHelper.SECOND_LINE;
import static com.matag.cards.properties.PowerToughness.powerToughness;
import static com.matag.game.turn.phases.main1.Main1Phase.M1;
import static com.matag.game.turn.phases.main2.Main2Phase.M2;
import static com.matag.player.PlayerType.OPPONENT;
import static com.matag.player.PlayerType.PLAYER;

public class CleanupTest extends AbstractApplicationTest {

  @Autowired
  private InitTestServiceDecorator initTestServiceDecorator;

  @Autowired
  private Cards cards;

  public void setupGame() {
    initTestServiceDecorator.setInitTestService(new CleanupTest.InitTestServiceForTest());
  }

  @Test
  public void cleanupWhenPassingTurns() {
    // Battlefields is
    browser.player1().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Huatli's Snubhorn")).hasDamage(1);
    browser.player1().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Huatli's Snubhorn")).isTapped();

    browser.player1().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Grazing Whiptail")).hasSummoningSickness();
    browser.player1().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Grazing Whiptail")).hasPowerAndToughness("4/5");

    browser.player2().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Air Elemental")).hasDamage(1);
    browser.player2().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Air Elemental")).isTapped();

    // Phase is
    browser.player1().getPhaseHelper().is(M1, PLAYER);

    // Player1 goes to M2
    browser.player1().getActionHelper().clickContinueAndExpectPhase(M2, PLAYER);

    // Player1 passes the turn
    browser.player1().getActionHelper().clickContinueAndExpectPhase(M1, OPPONENT);

    browser.player1().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Huatli's Snubhorn")).doesNotHaveDamage();
    browser.player1().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Huatli's Snubhorn")).isTapped();

    browser.player1().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Grazing Whiptail")).hasSummoningSickness();
    browser.player1().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Grazing Whiptail")).hasPowerAndToughness("3/4");

    browser.player2().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Air Elemental")).isNotTapped();
    browser.player2().getBattlefieldHelper(PLAYER, SECOND_LINE).getFirstCard(cards.get("Air Elemental")).doesNotHaveDamage();
  }

  static class InitTestServiceForTest extends InitTestService {
    @Override
    public void initGameStatus(GameStatus gameStatus) {
      // Current Player
      addCardToCurrentPlayerLibrary(gameStatus, cards.get("Island"));

      addCardToCurrentPlayerBattlefield(gameStatus, cards.get("Huatli's Snubhorn"));
      addCardToCurrentPlayerBattlefield(gameStatus, cards.get("Grazing Whiptail"));

      CardInstance huatlisShubhorn = gameStatus.getCurrentPlayer().getBattlefield().getCards().get(0);
      huatlisShubhorn.getModifiers().setTapped(true);
      huatlisShubhorn.getModifiers().dealDamage(1);

      CardInstance grazingWhiptail = gameStatus.getCurrentPlayer().getBattlefield().getCards().get(1);
      grazingWhiptail.getModifiers().setSummoningSickness(true);
      grazingWhiptail.getModifiers().getModifiersUntilEndOfTurn().addExtraPowerToughnessUntilEndOfTurn(powerToughness("1/1"));

      // Non Current Player
      addCardToNonCurrentPlayerLibrary(gameStatus, cards.get("Mountain"));

      addCardToNonCurrentPlayerBattlefield(gameStatus, cards.get("Air Elemental"));
      CardInstance airElemental = gameStatus.getNonCurrentPlayer().getBattlefield().getCards().get(0);
      airElemental.getModifiers().dealDamage(1);
      airElemental.getModifiers().setTapped(true);
    }
  }
}

File: src/test/kotlin/application/cleanup/CleanupTest.kt
package application.cleanup

import application.AbstractApplicationTest
import application.InitTestService
import application.browser.BattlefieldHelper
import application.cast.CastCreatureAlternativeCostTest.InitTestServiceForTest
import com.matag.cards.Cards
import com.matag.cards.properties.PowerToughness
import com.matag.game.adminclient.AdminClient
import com.matag.game.cardinstance.CardInstance
import com.matag.game.cardinstance.CardInstanceFactory
import com.matag.game.status.GameStatus
import com.matag.game.status.GameStatusRepository
import com.matag.game.turn.phases.main1.Main1Phase
import com.matag.game.turn.phases.main2.Main2Phase
import com.matag.player.PlayerType
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired

class CleanupTest(
    adminClient: AdminClient,
    gameStatusRepository: GameStatusRepository,
    cardInstanceFactory: CardInstanceFactory,
    cards: Cards,
    private var initService: InitTestService,
) : AbstractApplicationTest(adminClient, gameStatusRepository, cardInstanceFactory, cards) {

    override fun setupGame() {
        initService = InitTestServiceForTest(cardInstanceFactory, cards)
    }

    @Test
    fun cleanupWhenPassingTurns() {
        // Battlefields is
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards!!.get("Huatli's Snubhorn")).hasDamage(1)
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Huatli's Snubhorn")).isTapped()

        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Grazing Whiptail")).hasSummoningSickness()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Grazing Whiptail")).hasPowerAndToughness("4/5")

        browser.player2().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Air Elemental")).hasDamage(1)
        browser.player2().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Air Elemental")).isTapped()

        // Phase is
        browser.player1().getPhaseHelper().matches(Main1Phase.M1, PlayerType.PLAYER)

        // Player1 goes to M2
        browser.player1().getActionHelper().clickContinueAndExpectPhase(Main2Phase.M2, PlayerType.PLAYER)

        // Player1 passes the turn
        browser.player1().getActionHelper().clickContinueAndExpectPhase(Main1Phase.M1, PlayerType.OPPONENT)

        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Huatli's Snubhorn")).doesNotHaveDamage()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Huatli's Snubhorn")).isTapped()

        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Grazing Whiptail")).hasSummoningSickness()
        browser.player1().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Grazing Whiptail")).hasPowerAndToughness("3/4")

        browser.player2().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Air Elemental")).isNotTapped()
        browser.player2().getBattlefieldHelper(PlayerType.PLAYER, BattlefieldHelper.SECOND_LINE)
            .getFirstCard(cards.get("Air Elemental")).doesNotHaveDamage()
    }

    internal class InitTestServiceForTest(cardInstanceFactory: CardInstanceFactory, cards: Cards) : InitTestService(cardInstanceFactory, cards) {
        override fun initGameStatus(gameStatus: GameStatus) {
            // Current Player
            addCardToCurrentPlayerLibrary(gameStatus, cards.get("Island"))

            addCardToCurrentPlayerBattlefield(gameStatus, cards.get("Huatli's Snubhorn"))
            addCardToCurrentPlayerBattlefield(gameStatus, cards.get("Grazing Whiptail"))

            val huatlisShubhorn: CardInstance = gameStatus.currentPlayer.battlefield.cards.get(0)
            huatlisShubhorn.modifiers.isTapped = true
            huatlisShubhorn.modifiers.dealDamage(1)

            val grazingWhiptail: CardInstance = gameStatus.currentPlayer.battlefield.cards.get(1)
            grazingWhiptail.modifiers.isSummoningSickness = true
            grazingWhiptail.modifiers.modifiersUntilEndOfTurn
                .addExtraPowerToughnessUntilEndOfTurn(PowerToughness(1, 1))

            // Non Current Player
            addCardToNonCurrentPlayerLibrary(gameStatus, cards.get("Mountain"))

            addCardToNonCurrentPlayerBattlefield(gameStatus, cards.get("Air Elemental"))
            val airElemental: CardInstance = gameStatus.nonCurrentPlayer.battlefield.cards.get(0)
            airElemental.modifiers.dealDamage(1)
            airElemental.modifiers.isTapped = true
        }
    }
}
